File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/nodes/exec/TestingBatchExecNode.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.planner.plan.nodes.exec

import org.apache.flink.api.dag.Transformation
import org.apache.flink.table.api.TableConfig
import org.apache.flink.table.planner.calcite.{FlinkContextImpl, FlinkRelOptClusterFactory, FlinkRexBuilder, FlinkTypeFactory, FlinkTypeSystem}
import org.apache.flink.table.planner.delegation.BatchPlanner
import org.apache.flink.table.planner.plan.nodes.physical.batch.BatchPhysicalRel
import org.apache.flink.table.types.logical.{LogicalType, RowType}

import org.apache.calcite.plan.hep.{HepPlanner, HepProgram}
import org.apache.calcite.plan.{RelOptCluster, RelOptPlanner, RelTraitSet}
import org.apache.calcite.rel.AbstractRelNode
import org.apache.calcite.rel.`type`.RelDataTypeSystem

import java.util

/**
 * [[LegacyBatchExecNode]] for testing purpose.
 */
class TestingBatchExecNode
    extends AbstractRelNode(TestingBatchExecNode.cluster, TestingBatchExecNode.traitSet)
    with BatchPhysicalRel
    with LegacyBatchExecNode[BatchPlanner]  {

  val inputNodes: util.List[ExecNode[_]] =
    new util.ArrayList[ExecNode[_]]()
  val inputEdges: util.List[ExecEdge] = new util.ArrayList[ExecEdge]()

  def addInput(node: ExecNode[_]): Unit =
    addInput(node, ExecEdge.builder().build())

  def addInput(node: ExecNode[_], edge: ExecEdge): Unit = {
    inputNodes.add(node)
    inputEdges.add(edge)
  }

  override def getOutputType: LogicalType = RowType.of()

  override def getInputNodes: util.List[ExecNode[_]] = inputNodes

  override def getInputEdges: util.List[ExecEdge] = inputEdges

  override def replaceInputNode(
      ordinalInParent: Int,
      newInputNode: ExecNode[_]): Unit =
    inputNodes.set(ordinalInParent, newInputNode)

  override def getTraitSet: RelTraitSet = TestingBatchExecNode.traitSet

  override protected def translateToPlanInternal(
    planner: BatchPlanner): Transformation[BatchPlanner] = ???
}

object TestingBatchExecNode {
  val typeSystem: RelDataTypeSystem = new FlinkTypeSystem
  val typeFactory: FlinkTypeFactory = new FlinkTypeFactory(typeSystem)
  val context: FlinkContextImpl = new FlinkContextImpl(new TableConfig, null, null, null)
  val planner: RelOptPlanner = new HepPlanner(HepProgram.builder().build(), context)
  val cluster: RelOptCluster =
    FlinkRelOptClusterFactory.create(planner, new FlinkRexBuilder(typeFactory))
  val traitSet: RelTraitSet = RelTraitSet.createEmpty()
}

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/TestingBatchExecNode.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.planner.plan.nodes.exec;

import org.apache.flink.api.dag.Transformation;
import org.apache.flink.table.api.TableException;
import org.apache.flink.table.data.RowData;
import org.apache.flink.table.delegation.Planner;
import org.apache.flink.table.planner.plan.nodes.exec.batch.BatchExecNode;
import org.apache.flink.table.planner.plan.nodes.exec.visitor.ExecNodeVisitor;
import org.apache.flink.table.types.logical.LogicalType;
import org.apache.flink.table.types.logical.RowType;

import java.util.ArrayList;
import java.util.List;

/** {@link TestingBatchExecNode} for testing purpose. */
public class TestingBatchExecNode implements BatchExecNode<RowData> {

    private final List<ExecNode<?>> inputNodes;
    private final List<InputProperty> inputProperties;

    public TestingBatchExecNode() {
        this.inputNodes = new ArrayList<>();
        this.inputProperties = new ArrayList<>();
    }

    public void addInput(ExecNode<?> input) {
        addInput(input, InputProperty.DEFAULT);
    }

    public void addInput(ExecNode<?> input, InputProperty inputProperty) {
        inputNodes.add(input);
        inputProperties.add(inputProperty);
    }

    @Override
    public String getDescription() {
        return "TestingBatchExecNode";
    }

    @Override
    public LogicalType getOutputType() {
        return RowType.of();
    }

    @Override
    public List<ExecNode<?>> getInputNodes() {
        return inputNodes;
    }

    @Override
    public List<InputProperty> getInputProperties() {
        return inputProperties;
    }

    @Override
    public void replaceInputNode(int ordinalInParent, ExecNode<?> newInputNode) {
        inputNodes.set(ordinalInParent, newInputNode);
    }

    @Override
    public Transformation<RowData> translateToPlan(Planner planner) {
        throw new TableException("Unsupported operation.");
    }

    @Override
    public void accept(ExecNodeVisitor visitor) {
        visitor.visit(this);
    }
}
